### PR TITLE
Fixes #17811: fix nasty random generation failure

### PR DESCRIPTION
**Pull request by gpoblon**
https://issues.rudder.io/issues/17811

### CODE REVIEW

**Comment by gpoblon** [comment]
Generation of a technique in rudder-lang randomly failed for no obvious reason.

After investigation it appeared that this bug were exposed by a duplicate variable declaration located in the rudder-lang library.

The bug was hidden in a recursive function whose purpose was to generate the context tree (which is very important to register variables and their relative namespaces).

Duplicates weren't detected properly (recursion had flow flaws), and when they were (if the duplicate was the first entry of the context tree hashmap) it led to a crash.

Investigating this, I noticed another issue: namespaces weren't push right, ie:
```
sys.arch.linux
sys.arch.windows
```
Would led to
Context { arch { windows } }

instead of:
Context { arch { linux, windows } }

Now program behaves properly:
- warns the user when a duplicate is declared
- properly generates the tree regardless of its elements order
- properly pushes nested content
- added tests unit

File: rudder-lang/src/ast/context.rs [hunk 1]
```diff
@@ -129,10 +135,13 @@ impl<'src> VarContext<'src> {
         name: Token<'src>,
         value: Value<'src>,
     ) -> Result<()> {
+        // println!("new_variable");
+
         if let Some(ref mut existing_kind) = self.variables.get_mut(&name) {
             match existing_kind {
                 VarKind::Generic(ref mut existing_var) => {
-                    Self::push_new_variable(name, existing_var, &value)?
+                    // println!("SOME =>>>>> NAME = {}\n VALUE = {:#?}", name, value);
```
Comment (amousset) [inline rudder-lang/src/ast/context.rs:143]: this should be a proper debug or trace level log

**Comment by amousset** [comment]
Congratulations for finding this one!

File: rudder-lang/src/ast/context.rs (hunk 1 above)
Comment (gpoblon) [reply rudder-lang/src/ast/context.rs:143]: fixed

**Comment by gpoblon** [comment]
PR updated with a new commit

**Comment by gpoblon** [comment]
PR rebased

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**